### PR TITLE
Improve safety and logging of realm migration process

### LINKED ISSUE #16441: Add recovery support when realm cannot be opened
Currently if a realm file can't be opened (usually due to backwards migration), the game will not start.

Instead, the game should create a fresh realm file and allow itself to start, along with letting the user know what happened so they can report the issue if it was unexpected.

Marking as p0 since I'd want to get this ready before a release happens with #16428.

### PR DESCRIPTION
**Pull request by peppy**
This PR creates point-in-time backups of both EF and realm databases before both beatmap and score migration operations. To recover / retry the migration, the user just needs to remove the filename suffix on both.

Also adds more logging to the migration process, and adds back the final `client.db` deletion after the migration completes.

- [x] Depends on #16428

I probably wouldn't recommend reviewing this until the dependency is merged.

20220118 145128 (iTerm)

```csharp

[database] 2022-01-18 05:50:54 [verbose]: Beginning beatmaps migration to realm
[database] 2022-01-18 05:50:54 [verbose]: Found 99 beatmaps in EF
[database] 2022-01-18 05:50:54 [verbose]: Creating full EF database backup at client.before_beatmap_migration_1642485054.db
[database] 2022-01-18 05:50:54 [verbose]: Blocking realm operations.
[database] 2022-01-18 05:50:54 [verbose]: Creating full realm database backup at client.before_beatmap_migration_1642485054.realm
[database] 2022-01-18 05:50:54 [verbose]: Restoring realm operations.
[database] 2022-01-18 05:50:54 [verbose]: Successfully migrated 99 beatmaps to realm
[database] 2022-01-18 05:50:55 [verbose]: Beginning scores migration to realm
[database] 2022-01-18 05:50:55 [verbose]: No scores found to migrate
[database] 2022-01-18 05:50:55 [verbose]: Migration successful, deleting EF database

```

I did look at the potential of making a UI for this operation to give feedback, but it's not going to be a simple task (a lot of game-level components immediately start consuming realm on startup before we can show a UI). Going to wait to test against some large user databases before deciding if that's worth the effort of fixing.

Closes https://github.com/ppy/osu/issues/16441.

### CODE REVIEW

File: osu.Game/Database/EFToRealmMigrator.cs [hunk 1]
```diff
@@ -30,97 +33,119 @@ public EFToRealmMigrator(DatabaseContextFactory efContextFactory, RealmContextFa
...
 
             using (var realm = realmContextFactory.CreateContext())
-            using (var transaction = realm.BeginWrite())
             {
+                Logger.Log($"Found {existingBeatmapSets.Count} beatmaps in EF", LoggingTarget.Database);
+                string migration = $"before_beatmap_migration_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";
+                efContextFactory.CreateBackup($"client.{migration}.db");
+
                 // only migrate data if the realm database is empty.
                 // note that this cannot be written as: `realm.All<BeatmapInfo>().All(s => s.Protected)`, because realm does not support `.All()`.
```
Comment (bdach) [inline osu.Game/Database/EFToRealmMigrator.cs:77]: nit: outdated comment

```suggestion
```

File: osu.Game/Database/EFToRealmMigrator.cs [hunk 2]
```diff
@@ -151,62 +176,80 @@ private BeatmapMetadata getBestMetadata(EFBeatmapMetadata? beatmapMetadata, EFBe
...
 
             using (var realm = realmContextFactory.CreateContext())
-            using (var transaction = realm.BeginWrite())
             {
+                Logger.Log($"Found {existingScores.Count} scores in EF", LoggingTarget.Database);
+                string migration = $"before_score_migration_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";
+                efContextFactory.CreateBackup($"client.{migration}.db");
+
                 // only migrate data if the realm database is empty.
                 // note that this cannot be written as: `realm.All<ScoreInfo>().All(s => s.Protected)`, because realm does not support `.All()`.
```
Comment (bdach) [inline osu.Game/Database/EFToRealmMigrator.cs:202]: nit: this comment never fully made sense
```suggestion
```

File: osu.Game/Database/RealmContextFactory.cs [hunk 3]
```diff
@@ -361,6 +361,17 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
         private string? getRulesetShortNameFromLegacyID(long rulesetId) =>
             efContextFactory?.Get().RulesetInfo.FirstOrDefault(r => r.ID == rulesetId)?.ShortName;
 
+        public void CreateBackup(string filename)
+        {
+            using (BlockAllOperations())
+            {
+                Logger.Log($"Creating full realm database backup at {filename}", LoggingTarget.Database);
+                using (var source = storage.GetStream(Filename))
```
Comment (bdach) [inline osu.Game/Database/RealmContextFactory.cs:369]: kind of dangerous variable naming choices here, what with the `[fF]ilename` casing, but seems correct

File: osu.Game/Database/RealmContextFactory.cs [hunk 4]
```diff
@@ -374,17 +385,17 @@ public IDisposable BlockAllOperations()
...
-
-            Logger.Log(@"Blocking realm operations.", LoggingTarget.Database);
-
             try
             {
                 contextCreationLock.Wait();
 
                 lock (contextLock)
                 {
+                    if (!ThreadSafety.IsUpdateThread && context != null)
```
Comment (bdach) [inline osu.Game/Database/RealmContextFactory.cs:394]: The `context != null` check is to just make calling the migration from `OsuGameBase.load()` work, correct? Maybe it should be marked with a removal TODO at some point when it's no longer required?

Comment (peppy) [reply osu.Game/Database/RealmContextFactory.cs:394]: I'm not sure there will be a point it's not required. I think we want to be able to freely use realm during the initial load? Not sure.

Comment (bdach) [reply osu.Game/Database/RealmContextFactory.cs:394]: I suppose there shouldn't be much of an issue with that, yeah.

File: osu.Game/Database/RealmContextFactory.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Database/RealmContextFactory.cs:369]: yeah, although it's a nit, I think it's safer to rename the parameter according to what it holds (i.e. `backupFilename`).

**Comment by bdach** [comment]
A few other loose thoughts:

* I'm curious if there was a reason for taking snapshots of both realm and EF before score and beatmap migrations individually rather than just the one snapshot of both before everything?
* I've also [gone back to the discord discussion from earlier](https://discord.com/channels/188630481301012481/188630652340404224/932945257774071818) and having given it more thought I think that I still prefer the approach taken here (make backups, migrate everything, clean out EF). While I can see that leaving the old client db intact would be more convenient for both devs and users who want to downgrade, I'm worried that especially users may end up breaking their file store without even being aware about it due to going back and forth, and we'll have to keep chasing weird reports of databases being in weird states that should be otherwise impossible. To a degree I'm still expecting there to be issues, but with the current method at least I'd expect them to be loud and very distinctive.
 
  I could however agree with [leaving the old EF database intact on debug builds](https://discord.com/channels/188630481301012481/188630652340404224/932947201922052096) as a compromise. Not sure what @smoogipoo's thoughts on that are (and it kind of works for us because we're here every day, but may not work for occasional contributors).

**Comment by peppy** [comment]
> A few other loose thoughts:
> 
> 
> 
> * I'm curious if there was a reason for taking snapshots of both realm and EF before score and beatmap migrations individually rather than just the one snapshot of both before everything?

The rationale was that one could silently fail and it could help in debugging, but that's probably not good enough reason. It made more sense when deletions were happening in EF per row/migration. One backup should be enough now.

File: osu.Game/Database/EFToRealmMigrator.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Database/EFToRealmMigrator.cs:77]: Due to the `BeatmapInfo` instead of `BeatmapSetInfo`?

**Comment by peppy** [comment]
I've added a notification that shows up only for debug builds to potentially ease concerns.

20220119 103303 (osu!)

File: osu.Game/Database/RealmContextFactory.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Database/RealmContextFactory.cs:369]: I've renamed the variable.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**